### PR TITLE
Feature/Article pinning

### LINKED ISSUE #4: Article pinning
Copied from alpha-banana-api (dutch):
Het zou handig zijn als je een artikel (tijdelijk) verder omhoog kan zetten. Dit zouden we kunnen doen met een prioriteits flag. We zouden 2 fields kunnen toevoegen: priority en priority_untill. Dan zouden we een sort kunnen maken die het artikel met de hoogste priority die nog valid is (current_time < priority_untill) bovenaan zet.

@Matthijsy: Update van de vergadering van 26-03. Het idee is om 1 artikel te kunnen pinnen. Het idee is dan om een pinned_untill veld toe te voegen. Maar 1 artikel mag dan tegelijk gepinned zijn.

### PR DESCRIPTION
**Pull request by guidojw**
This PR adds a "pinned" field to the article model that the UI can use to indicate which article is pinned and should thus be shown first. The field defaults to false and only one article record can be pinned concurrently. An article has to be unpinned manually.

This PR resolves #4

### CODE REVIEW

File: app/models/article.rb [hunk 1]
```diff
@@ -21,4 +25,13 @@ def owners
       [author]
     end
   end
+
+  private
+
+  def only_one_pinned_article
+    return unless pinned?
+
+    matches = Article.pinned.where('id != ?', id)
```
Comment (Matthijsy) [inline app/models/article.rb:34]: `Article.pinned.where.not(id: id)` is also valid I think, and a bit more clean

File: db/schema.rb [hunk 2]
```diff
@@ -267,6 +268,7 @@
     t.datetime "deleted_at"
     t.bigint "moderator_group_id"
     t.boolean "smtp_enabled", default: false
+    t.datetime "last_received_at"
```
Comment (Matthijsy) [inline db/schema.rb]: This is not related to this PR. Could you remove it? And check if the migration is valid? Run the following:
1. `git checkout staging`
1. `rails db:reset`
1. `git checkout <branch_name>`
1. `rails db:migrate`

This checks if your migrations works from staging as well

File: spec/models/article_spec.rb [hunk 3]
```diff
@@ -52,6 +58,32 @@
...
   end
 
+  describe '#pinned' do
+    context 'when other already pinned' do
+      subject(:article) { FactoryBot.create(:article, pinned: false) }
+
+      before do
+        FactoryBot.create(:article, pinned: true)
+        article.pinned = true
+        article.valid?
```
Comment (Matthijsy) [inline spec/models/article_spec.rb:68]: You don't have to run this

File: spec/models/article_spec.rb [hunk 4]
```diff
@@ -52,6 +58,32 @@
     it { expect(described_class.count - described_class.publicly_visible.count).to be 1 }
   end
 
+  describe '#pinned' do
+    context 'when other already pinned' do
+      subject(:article) { FactoryBot.create(:article, pinned: false) }
```
Comment (Matthijsy) [inline spec/models/article_spec.rb:63]: you can also do this `subject(:article) { FactoryBot.build(:article, pinned: true) }`. Then it is not inserted into the database, but you can validate it. If you then just insert another one in the database (like you did below) it will become invalid

File: db/schema.rb (hunk 2 above)
Comment (guidojw) [reply db/schema.rb]: Yup that was fixed by merging staging into the branch.
I've ran those commands and it migrated successfully

File: app/resources/v1/article_resource.rb [hunk 5]
```diff
@@ -30,8 +30,15 @@ def fetchable_fields
...
-    %i[title content publicly_visible group cover_photo]
+  def self.creatable_fields(context)
+    attributes = %i[title content publicly_visible group cover_photo]
+
+    attributes += [:pinned] if update_permission?(context)
+    attributes
+  end
+
+  def self.update_permission?(context)
+    context[:user]&.permission?(:update, _model_class)
```
Comment (Matthijsy) [inline app/resources/v1/article_resource.rb:41]: Can't you use `update_permission?`? Not sure if that is available at the `self` context, but just wanted to check

File: app/resources/v1/article_resource.rb [hunk 6]
```diff
@@ -30,8 +30,15 @@ def fetchable_fields
     super - [:cover_photo]
   end
 
-  def self.creatable_fields(_context)
-    %i[title content publicly_visible group cover_photo]
+  def self.creatable_fields(context)
+    attributes = %i[title content publicly_visible group cover_photo]
+
+    attributes += [:pinned] if update_permission?(context)
```
Comment (Matthijsy) [inline app/resources/v1/article_resource.rb:36]: great! Could you write a little spec for this? See for example `/spec/resources/v1/user_resource_spec.rb`

File: app/resources/v1/article_resource.rb (hunk 5 above)
Comment (guidojw) [reply app/resources/v1/article_resource.rb:41]: I tried to but it was indeed not available

Comment (guidojw) [reply app/resources/v1/article_resource.rb:41]: Is this method correct by the way? I wasn't sure what to use to see if someone has article.update permissions.

Comment (Matthijsy) [reply app/resources/v1/article_resource.rb:41]: Yes it is

Comment (guidojw) [reply app/resources/v1/article_resource.rb:41]: This is also done in the user resource btw (https://github.com/csvalpha/amber-api/blob/7c244de1a2ad1540794e99214f293ec0dbabc874/app/resources/v1/user_resource.rb#L115-L117)
